spark/dl/src/main/scala/com/intel/analytics/bigdl/optim/ParallelOptimizer.scala

```diff
@@ -0,0 +1,1053 @@
+            Iterator.empty
+          }).collect
+        }
+        validate(
+          validationTrigger,
+          validationDataSet,
+          validationMethods,
+          coresPerNode,
+          models,
+          driverState,
+          validationSummary,
+          _header
+        )
+
+        trainSummary.foreach { summary =>
+          saveSummary(
+            summary,
+            models,
+            driverState,
+            trainingModel
+          )
+        }
+
+        checkpoint(
+          cacheTrigger,
+          cachePath,
+          isOverWrite,
+          wallClockTime,
+          models,
+          driverState,
+          optimMethods,
+          trainingModel
+        )
+      } else {
+        logger.info(s"Warning! Not enough training samples were successfully processed in this " +
+          s"iteration due to some slow tasks. The gradients computed in this iteration will be " +
+          s"discarded. Only $numFinishedModelUpdates/$driverSubModelNum threads successfully " +
+          s"completed training.")
+      }
+    }
+  }
+
+  private def updateLayerParameters[T: ClassTag](module: Module[T]): Unit = {
+    module.updateParameter
+    if (module.isInstanceOf[Container[_, _, T]]) {
+      module.asInstanceOf[Container[_, _, T]].modules.foreach(sub => {
+        updateLayerParameters(sub)
+      })
+    }
+  }
+
+  /**
+   * Create checkpoint.
+   *
+   * @param cacheTrigger cache trigger
+   * @param cachePath cache path
+   * @param isOverWrite whether over write
+   * @param wallClockTime wall clock time
+   * @param models cached models
+   * @param state state table
+   */
+  private def checkpoint[T: ClassTag](
+    cacheTrigger: Option[Trigger],
+    cachePath: Option[String],
+    isOverWrite: Boolean,
+    wallClockTime: Long,
+    models: RDD[Cache[T]],
+    state: Table,
+    optimMethods: Map[String, OptimMethod[T]],
+    trainingModel: Module[T])(implicit ev: TensorNumeric[T]): Unit = {
+    cacheTrigger.foreach { trigger =>
+      cachePath.foreach { path =>
+        if (trigger(state)) {
+          saveModel(getModel(models, trainingModel), cachePath, isOverWrite,
+            s".${state[Int]("neval")}")
+          logger.info(s"[Wall Clock ${wallClockTime / 1e9}s] Save model to $path")
+          optimMethods.foreach{case (name, optimMethod) =>
+            optimMethod.state.update("epoch", state[Int]("epoch"))
+            optimMethod.state.update("neval", state[Int]("neval"))
+            saveOptimMethod(optimMethod, cachePath, isOverWrite, s"-$name.${state[Int]("neval")}")
+            logger.info(s"[Wall Clock ${wallClockTime / 1e9}s] Save optimMethod " +
+              s"${optimMethod} to $path")
+          }
+        }
+      }
+    }
+  }
+
+  /**
+   * Save train summaries.
+   *
+   * @param trainSummary train logger
+   * @param models cached models
+   * @param driverState driver state
+   */
+  private def saveSummary[T: ClassTag](
```

Comment:
can we share this method with the original optimizer?

Reply:
yep, makes sense